proxylib: Enable parser tests in their own package.

jrajahalme opened the pull request:
Moving most of the functionality not related to the shared library interface to the proxylib package enables testing in the parser packages. The actual test parsers are still tested in the main package to exercise the shared library interface functions.

This commit moves the Cassandra unit tests to the cassandra package using the new testing tools in the proxylib package. The tests are somewhat more readable doe to reduced boilerplate.

Accesslog client is simplified now that the path of an existing client never changes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5735)
<!-- Reviewable:end -->

raybejjani commented on proxylib/proxylib.go:
```diff
@@ -63,61 +58,14 @@ func OnNewConnection(instanceId uint64, proto string, connectionId uint64, ingre
-			bytes -= rem
-			input = input[1:] // may result in an empty slice
-		}
+	res, conn := NewConnection(instance, strcpy(proto), connectionId, ingress, srcId, dstId, strcpy(srcAddr), strcpy(dstAddr), strcpy(policyName), origBuf, replyBuf)
```
I don't follow what strcpy does here (but it was already in the code so not relevant to this PR). In particular, aren't there cgo functions to make copies such as this? When the string is recreated I think go will reuse the prior instance (strings are interned).

jrajahalme replied:
Cgo allow strings and slices of trivial types to be passed from C to Go and Cgo defines the C-types for string and slice headers. I my experience Cgo does not (thankfully) copy the underlying data. Cgo documentation explicitly states that C can't hold references to Go memory (due to GC). It is not so strict in the other direction, but generally the C side needs to keep any memory passed to Go valid, if references are retained longer than for the duration of the function call.

In this specific case we are passing strings from C to Go. The problem is that I have no provisions on the calling side to keep any of the underlying memory around longer than the duration of the function call, so the data might be in the stack, etc.

According to Go documentation the underlying data is copied whenever a string is cast/sliced to a slice of bytes, and the other way around. My definition of `strcpy()` uses this and first casts  the string as a slice of bytes, and then creates a string using those bytes. In my experience this works. However, maybe it would be safer (long term) for `strcpy()` to use `copy()` explicitly for the first copy, and then cast it back to string?

raybejjani replied:
Oh, I meant using `func C.GoString(*C.char) string` which apparently makes a copy:
 "A few special functions convert between Go and C types by making copies of the data" (https://golang.org/cmd/cgo/)
but I actually don't fully trust that ;) What you have does the same thing.
I had read the strcpy docstring as making a new copy every time but, reading it again, that's clearly not the expectation (just a copy from the C). Anyway, not a big deal!

raybejjani commented on proxylib/proxylib/connection.go:
```diff
@@ -59,6 +62,117 @@ type Connection struct {
+		// Recover from any possible parser datapath panics
+		if r := recover(); r != nil {
+			// Log the Panic into accesslog
+			connection.Log(cilium.EntryType_Denied,
```
Since this happens in a panic, is is possible to try and lock an already locked object? Looking at `connection.Log` it doesn't seem like it.

jrajahalme replied:
Good point! Would have a problem if `accesslog` `Log()` would panic and did not release locks. Typically, when an access log unix domain socket is already connected, `Log()` would not lock at all, but even when it does, it defers the unlock, so any panic therein would release the lock before the `Log()` here is executed.

raybejjani commented on proxylib/proxylib/types.go:
```diff
@@ -21,12 +21,8 @@ import "C"
-	op      uint32
-	n_bytes uint32 // The number of bytes of data the operation 'op' applies to.
-}
+type OpType int64
```
It might be strictly worse, but could these use `C.uint` `C.long` directly? In theory that would follow any architecture changes on the C side but those are probably fixed width as well.

jrajahalme replied:
Generally, you can't use `C.` types without using Cgo, and I have wanted to limit Go-files using Cgo tp the minimum. On reason for this is that any types defined if different Go packages, even if defined identically, are not the same type. 

Also, in a library interface it is safer to used explicitly sized types rather than assume both sides agree on a size of a type.

raybejjani commented on proxylib/proxylib/connection.go:
```diff
@@ -59,6 +62,117 @@ type Connection struct {
+	return input
+}
+
+func (connection *Connection) OnData(reply, endStream bool, data *[][]byte, filterOps *[][2]int64) (res FilterResult) {
```
I'm loth to hide that this is a 2 element array, but it might be worthwhile making it a type and giving it accessors for `op` and `bytes`. I'm mostly concerned that we don't have any documentation on what's stored in each element.

jrajahalme replied:
I did have a `C.FilterOp` type there, but hit difficulty due to the same `types.h` file included from `main` and `proxylib` packages defining their own definitions of that type. I have not fould examples of using C types of another package, but maybe it would be possible to include `types.h` only in the `proxylib/proxylib` package and then use `proxylib.C.FilterOp` etc. here instead?

jrajahalme replied:
Tried this again, but I can't find a way to use a C type defined in another package. Using `package.` prefix does not work, and without it (using `*[]C.FilterOp` I get:
```
./proxylib.go:103: cannot use filterOps (type *[]_Ctype_struct___0) as type *[]proxylib._Ctype_struct___0 in argument to connection.OnData
```
and with it (`*[]proxylib.C.FilterOp`) I get:
```
./proxylib.go:94:96: missing ',' in parameter list
./proxylib.go:96:7: expected type, found '.'
./proxylib.go:97:17: expected type, found ':='
./proxylib.go:98:15: missing ',' in parameter list
./proxylib.go:99:2: expected declaration, found 'if'
./proxylib.go:122:2: expected declaration, found 'for'
```
Go compiler seems to get confused about the two dots in the type name?

raybejjani replied:
Hmm, yeah, you are right. I think the `C` is local to the package the import happens in, despite it being uppercase. My only concern was getting the indexes mixed. A comment somewhere is probably good enough (I guess you could also make named variables for each index but that feels a little yucky).

raybejjani reviewed (approved):
All my comments are about the code in general, so nothing blocking this PR.

aanm commented on proxylib/cassandra/cassandraparser_test.go:
```diff
@@ -0,0 +1,282 @@
+	"github.com/cilium/cilium/proxylib/proxylib"
+	"github.com/cilium/cilium/proxylib/test"
+
+	// log "github.com/sirupsen/logrus"
```
Leftover?

jrajahalme replied:
Yes and no. Toggling on the debug log level when developing new tests requires these lines. My preference is to keep these as comments rather than re-insert them every time new tests are added.

aanm commented on proxylib/cassandra/cassandraparser_test.go:
```diff
@@ -0,0 +1,282 @@
+	"encoding/hex"
+	"testing"
+
+	// "github.com/cilium/cilium/pkg/logging"
```
Leftover?

jrajahalme replied:
See above.

aanm commented on proxylib/cassandra/cassandraparser_test.go:
```diff
@@ -0,0 +1,282 @@
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) {
+	// logging.ToggleDebugLogs(true)
```
Leftover? you can use `-v` in go check.

for example if I run `go test -check.f "Test.*addCiliumNetworkPolicyV2" -v` inside `daemon/` it will only test the test `addCiliumNetworkPolicyV2`.

jrajahalme replied:
`-v` does not trigger debug logging though. Would be handy if we could simply issue `-debug`.

aanm commented on proxylib/accesslog/client.go:
```diff
@@ -28,23 +28,24 @@ import (
-	mutex     lock.Mutex // Used to protect opening the connection
+	connected uint32 // Accessed atomically without locking
 	path      string
+	mutex     lock.Mutex     // Used to protect opening the connection
```
If this is only protecting the `unsafe.Pointer` bellow consider using the https://golang.org/pkg/sync/atomic/#LoadPointer and https://golang.org/pkg/sync/atomic/#StorePointer and user a wrapper function like `GetConn` `SetConn`

jrajahalme replied:
We already use `LoadPointer` and `StorePointer`, but I also wanted to make sure we do not attempt to open two connections to the access log server at the same time. So there is a larger critical section the mutex protects, consisting of
- read pointer value
- connect to access log server unix domain socket
- store pointer value
- store a boolean indicating the that connection was successfully opened

The above critical section needs to be entered only after send failures, otherwise the access logging takes no locks even when the connection is shared between multiple threads.

aanm commented on proxylib/proxylib/connection.go:
```diff
@@ -59,6 +62,117 @@ type Connection struct {
+	// Find the parser for the proto
+	parserFactory := GetParserFactory(proto)
+	if parserFactory == nil {
+		return UNKNOWN_PARSER, nil
```
Why isn't `NewConnection` returning an error for these errors?

jrajahalme replied:
Did not think of it, as the result is returned via the shared library interface as-is (not as a go error). However, it was simple enough for the `Result` type to implement the error interface so I changed that. Thanks!

aanm commented on proxylib/proxylib/test_util.go:
```diff
@@ -0,0 +1,120 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxylib
```
Why isn't this file called `util_test.go`?

jrajahalme replied:
It contains no tests and apparently `*_test.go` files are not imported when the package is imported. If I rename the file as suggested, and try to run the cassandra unit tests, I get:
```
vagrant@runtime1:~/go/src/github.com/cilium/cilium/proxylib/cassandra$ go test -test.v -check.vv
# github.com/cilium/cilium/proxylib/cassandra
./cassandraparser_test.go:80:15: s.ins.CheckNewConnectionOK undefined (type *proxylib.Instance has no field or method CheckNewConnectionOK)
./cassandraparser_test.go:87:7: s.ins.CheckInsertPolicyText undefined (type *proxylib.Instance has no field or method CheckInsertPolicyText)
./cassandraparser_test.go:108:15: s.ins.CheckNewConnectionOK undefined (type *proxylib.Instance has no field or method CheckNewConnectionOK)
./cassandraparser_test.go:118:7: s.ins.CheckInsertPolicyText undefined (type *proxylib.Instance has no field or method CheckInsertPolicyText)
./cassandraparser_test.go:139:15: s.ins.CheckNewConnectionOK undefined (type *proxylib.Instance has no field or method CheckNewConnectionOK)
./cassandraparser_test.go:146:7: s.ins.CheckInsertPolicyText undefined (type *proxylib.Instance has no field or method CheckInsertPolicyText)
./cassandraparser_test.go:167:15: s.ins.CheckNewConnectionOK undefined (type *proxylib.Instance has no field or method CheckNewConnectionOK)
./cassandraparser_test.go:175:7: s.ins.CheckInsertPolicyText undefined (type *proxylib.Instance has no field or method CheckInsertPolicyText)
./cassandraparser_test.go:196:15: s.ins.CheckNewConnectionOK undefined (type *proxylib.Instance has no field or method CheckNewConnectionOK)
./cassandraparser_test.go:204:7: s.ins.CheckInsertPolicyText undefined (type *proxylib.Instance has no field or method CheckInsertPolicyText)
./cassandraparser_test.go:204:7: too many errors
FAIL	github.com/cilium/cilium/proxylib/cassandra [build failed]
```

aanm commented on proxylib/proxylib/types.go:
```diff
@@ -21,12 +21,8 @@ import "C"
-	op      uint32
-	n_bytes uint32 // The number of bytes of data the operation 'op' applies to.
-}
+type OpType int64
```
Why aren't these `uint64`?

jrajahalme replied:
C types cannot be used in multiple packages (here `main` and `proxylib`; see my responses to Ray's questions for more context), so I had to change the Go-representation of `FilterOp` type to an array instead of a struct. Go length types are (64-bit) signed, so I went for int64 for the array type.

aanm commented on proxylib/proxylib/types.go:
```diff
@@ -21,12 +21,8 @@ import "C"
-	n_bytes uint32 // The number of bytes of data the operation 'op' applies to.
-}
+type OpType int64
+type OpError int64
```
Why aren't these `uint64`?

jrajahalme replied:
See above for `OpType`.

Same applies for `OpError`, but additionally it consumes the same space as `n_bytes` does for non-error operations.

aanm commented on proxylib/testhelpers.go:
```diff
@@ -45,22 +45,22 @@ func checkConnectionCount(t *testing.T, expConns int) {
 }
 
-func checkConnections(t *testing.T, res, expected Result, expConns int) {
+func checkConnections(t *testing.T, res, expected FilterResult, expConns int) {
```
why aren't this file called `something_test.go`

jrajahalme replied:
`*_test.go` files can't use Cgo, and we want to test the exported functions that before this patch used C types. Now that we are not using C types here anymore this file can be renamed, thanks!

aanm reviewed (changes requested):
@jrajahalme for the test helpers files, did you consider having a package called "something_test"? For example https://golang.org/pkg/net/http/httptest/

jrajahalme committed:
proxylib: Enable parser tests in their own package

Moving most of the functionality not related to the shared library
interface to the proxylib package enables testing in the parser
packages. The actual test parsers are still tested in the main package
to exercise the shared library interface functions.

This commit moves the Cassandra unit tests to the cassandra package
using the new testing tools in the proxylib package. The tests are
somewhat more readable doe to reduced boilerplate.

Accesslog client is simplified now that the path of an existing client
never changes.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
